app/src/citations.py

```diff
@@ -43,78 +61,81 @@ def create_prompt_context(chunks: Sequence[Chunk]) -> str:
+    Remap the ChunkWithSubsection.id value to be the user-friendly citation number for that citation.
+    E.g., if `subsections` is a list with five entries, and `response` is a string like
     "Example (citation-3)(citation-1), another example (citation-1).", then this function will return
-    {context[3]: 1, context[1]: 2}; citations referencing context[3] should be shown to the user as "1" and
-    citations referencing context[1] should be shown as "2".
+    [ChunkWithSubsection(id="1",...), ChunkWithSubsection(id="2",...)] so that
+    citations referencing '(citation-3)' can be shown to the user with id="1" and
+    citations referencing '(citation-1)' can be shown with id="2".
 
     """
     citation_indices = re.findall(CITATION_PATTERN, response)
 
-    # Maintain order and avoid duplicates by using a list and tracking seen indices
-    citations: dict[ChunkWithSubsection, int] = {}
-    citation_number = 1
-
-    for index in citation_indices:
+    # Avoid duplicates while maintaining citation order
+    deduped_indices = []
+    for x in citation_indices:
+        if x not in deduped_indices:
+            deduped_indices.append(x)
+
+    # Create a lookup map from original citation id to the corresponding ChunkWithSubsection
+    citation_map = {c.id: c for c in subsections}
+    # Factory to create new consecutive citation ids
+    factory = CitationIdFactory(start=1, prefix="")
+    # Citations to be returned; note that uncited subsections are not included
+    citations: list[ChunkWithSubsection] = []
+    for citation_id in deduped_indices:
+        # Since deduped_indices are generated by the LLM, it's possible for it to hallucinate a citation_id,
+        # so check that each citation_id exists.
+        if citation_id in citation_map:
+            # Add a copy of the subsection with the id replaced by a new consecutive citation number
+            citation = citation_map[citation_id]
+            citations.append(factory.create_citation(citation.chunk, citation.subsection))
+    return citations
 
-        index = int(index)  # Convert string to integer index
-        # Check that index is in-bounds, since these are generated by the LLM,
-        # so it's possible for it to hallucinate index numbers that don't exist
-        if 0 <= index < len(context) and context[index] not in citations:
-            citations[context[index]] = citation_number
-            citation_number += 1
 
-    return citations
+# TODO: move this to format.py
+def reify_citations(response: str, subsections: Sequence[ChunkWithSubsection]) -> str:
+    remapped_citations = remap_citation_ids(subsections, response)
+    return add_citation_links(response, remapped_citations)
 
 
-def reify_citations(response: str, chunks: list[Chunk]) -> str:
+# TODO: move this to format.py since this is UI logic
+def add_citation_links(response: str, remapped_citations: Sequence[ChunkWithSubsection]) -> str:
```

Comment:
Thoughts about moving `reify_citations()` and `add_citation_links()` to `format.py` since this is UI logic?